src/schema_traverse.rs

```diff
@@ -0,0 +1,191 @@
+use crate::tools::py_err;
+use pyo3::exceptions::{PyException, PyKeyError};
+use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyList, PySet, PyString, PyTuple};
+use pyo3::{create_exception, intern, Bound, PyResult};
+use std::collections::HashSet;
+
+create_exception!(pydantic_core._pydantic_core, GatherInvalidDefinitionError, PyException);
+
+macro_rules! get {
+    ($dict: expr, $key: expr) => {
+        $dict.get_item(intern!($dict.py(), $key))?
+    };
+}
+
+macro_rules! traverse_key_fn {
+    ($key: expr, $func: expr, $dict: expr, $ctx: expr) => {{
+        if let Some(v) = get!($dict, $key) {
+            $func(v.downcast_exact()?, $ctx)?
+        }
+    }};
+}
+
+macro_rules! traverse {
+    ($($key:expr => $func:expr),*; $dict: expr, $ctx: expr) => {{
+        $(traverse_key_fn!($key, $func, $dict, $ctx);)*
+        traverse_key_fn!("serialization", gather_schema, $dict, $ctx);
+        gather_meta($dict, $ctx)
+    }}
+}
+
+macro_rules! defaultdict_list_append {
+    ($dict: expr, $key: expr, $value: expr) => {{
+        match $dict.get_item($key)? {
+            None => {
+                let list = PyList::empty_bound($dict.py());
+                list.append($value)?;
```

Comment:
Potentially a touch faster:

```suggestion
                let list = PyList::new_bound($dict.py(), [value]);
```

Reply:
Done